presto-native-execution/presto_cpp/main/Announcer.cpp

```diff
@@ -166,10 +174,24 @@ void Announcer::scheduleNext() {
   if (stopped_) {
     return;
   }
+
+  auto timeToSleepMs = frequencyMsMax_;
+  if (failedAttempts_ > 0) {
+    auto rng = folly::ThreadLocalPRNG();
+    timeToSleepMs =
+        folly::futures::detail::retryingJitteredExponentialBackoffDur(
+            failedAttempts_,
+            std::chrono::milliseconds(frequencyMsMin_),
+            std::chrono::milliseconds(frequencyMsMax_),
+            jitterParam,
+            rng)
+            .count();
+  }
+
   eventBaseThread_.getEventBase()->scheduleAt(
       [this]() { return makeAnnouncement(); },
       std::chrono::steady_clock::now() +
-          std::chrono::milliseconds(frequencyMs_));
+          std::chrono::milliseconds(timeToSleepMs));
```

Comment:
There should also be a jitter If announcement is always successful. In that case, we can do something like `timeToSleepMs + rand(2000) - 1000`

Reply:
In high level, when a worker makes a successful announcement, it should not schedule the next announcement very quickly, but after some specific period of time which should ideally include some jitter (as you suggested). However, when the announcement fails, worker should try the next attempt quickly with exponential back-off. And also, this wait can't be infinitely long, so capped by max delay (which is today's hard-coded frequency ms). Since worker does not know when coordinator will come back, it will need to talk indefinitely. The issue is indefinite talking is - if there are too many failed attempts then all workers will get synchronized towards the frequencyMsMax (somewhat today's behavior). On the other side, if frequencyMsMax is too big, then if coordinator comes back after a number of failed attempts, worker registration will be late due to exponential back off. That will hurt routing and caching etc.

Hence a reasonable trade off is:
 - Add some jitter in success
 - Back-off the retry exponentially where initial failed attempts try to talk quickly hopoing coordinator is back soon and registration can be done quickly (this is improvement from today's hard coded timing, because today worker need to wait a lot regardless of the failed attempts)
 - We need to keep the announcement indefinitely (for both success and failure) to keep giving heartbead to coordiantor.

Reply:
I think it might be good to add a small amount of jitter even in successful case to avoid any possible synchronized announcements from all the workers especially in batch cluster with a lot of workers. thanks!